src/utils/auth.ts

```diff
@@ -0,0 +1,239 @@
+    const initPromises: Promise<void>[] = [];
+
+    initPromises.push(
+      (async (publicClient: PublicClientApplication) => {
+        try {
+          const response = await publicClient.handleRedirectPromise();
+          this.handleLoginResponse(response, null);
+        } catch (err) {
+          this.handleLoginResponse(null, err);
+        }
+      })(this.publicClient)
+    );
+
+    const account = this.getActiveAccount();
+
+    if (account) {
+      console.log('found cached account', account);
+      this._account = account;
+    }
+
+    initPromises.push(
+      (async () => {
+        if (this.loginPromise) await this.loginPromise;
+        this.loginPromise = this.login(loginRequest);
+        await this.loginPromise;
+      })()
+    );
+
+    this.initPromise = Promise.all(initPromises);
+  }
+
+  private handleLoginResponse(
+    response: AuthenticationResult | null,
+    err: Error | null
+  ) {
+    if (err) {
+      console.error('login error', err);
+      return;
+    }
+
+    console.log('login result', response);
+
+    const account = this.getActiveAccount();
+
+    if (response) {
+      console.log('login success, has response', account);
+      this._account = account;
+      this._token = response;
+    } else {
+      if (account) {
+        console.log('has accounts', account);
+        this._account = account;
+      }
+    }
+  }
+
+  private handleApiTokenResponse(
+    response: AuthenticationResult | null,
+    err: Error | null
+  ) {
+    if (err) {
+      console.error('api token error', err);
+      return;
+    }
+
+    console.log('api token result', response);
+
+    if (response) {
+      if (!response.accessToken) {
+        console.error(
+          'api token failed, got response but no accessToken field'
+        );
+      }
+    } else {
+      console.error('api token failed, null response but no error');
+    }
+  }
+
+  private getActiveAccount(): AccountInfo | null {
+    const allAccounts = this.publicClient.getAllAccounts();
+
+    if (allAccounts?.length) return allAccounts[0];
+
+    return null;
+  }
+
+  private async login(loginRequest: RedirectRequest) {
+    if (this.isAuthenticated) {
+      return;
+    }
+
+    try {
+      let request = this.createRequestWithLoginHint(loginRequest);
+  
+      if (request instanceof Error) {
+        throw new InteractionRequiredAuthError('cant_create_request');
+      }
+
+      const response = await this.publicClient.ssoSilent(request);
+      this.handleLoginResponse(response, null);
+    } catch (err) {
+      if (err instanceof InteractionRequiredAuthError) {
+        try {
+          await this.publicClient.loginRedirect(loginRequest);
+        } catch (err) {
+          if (err instanceof BrowserAuthError && (err as BrowserAuthError).errorCode === BrowserAuthErrorMessage.interactionInProgress.code) {
+            console.warn('Tried to issue login via redirect while interaction was in progress, this should be fixed in the future');
```

Comment:
This happens when the page load post-redirect _I think_, probably should be able to detect that we're in an interection, so will see if I'm missing something here if I get time. If not, no biggie, it still works